src/rgw/rgw_string.h

```diff
@@ -233,4 +238,167 @@ extern bool match_wildcards(boost::string_view pattern,
                             boost::string_view input,
                             uint32_t flags = 0);
 
+inline void rgw_escape_str(const std::string& s, const char esc_char, const char special_char, std::string& dest)
+{
+ dest.reserve(s.size() * 2);
+ 
+ auto dest_iter = begin(dest);
+ 
+ for (const char c : s) {
+    if(esc_char == c || special_char == c) {
+        *dest_iter++ = esc_char;
+    }
+    
+    *dest_iter++ = c;
+ }
+}
+
+inline void rgw_escape_str(const std::string& s, const char esc_char, const char special_char, std::string *dest_ptr)
+{
+ return rgw_escape_str(s, esc_char, special_char, *dest_ptr);
+}
+
+inline std::string rgw_escape_str(const std::string& s, const char esc_char, const char special_char)
+{
+ std::string result;
+ rgw_escape_str(s, esc_char, special_char, result);
+ return result;
+}
+
+inline ssize_t rgw_unescape_str(const std::string& src, const ssize_t ofs,
+                       const char esc_char, const char special_char,
+                       std::string& dest)
+{
+  dest.reserve(src.size());
+  auto destp = begin(dest);
+
+  bool esc = false;
+
+  for (size_t i = ofs; i < src.size(); i++) {
+    char c = src[i];
+    if (!esc && c == esc_char) {
+      esc = true;
+      continue;
+    }
+    if (!esc && c == special_char) {
+      dest.resize(i);
+      return dest.size();
+    }
+    *destp++ = c;
+    esc = false;
+  }
+
+  return std::string::npos;
+}
+
+inline ssize_t rgw_unescape_str(const std::string& s, const ssize_t ofs,
+                       const char esc_char, const char special_char,
+                       std::string *dest_ptr)
+{
+  return rgw_unescape_str(s, ofs, esc_char, special_char, *dest_ptr);
+}
+
+class CephContext;
```

Comment:
I never cared for this, since I feel like it makes the code more fragile and breaks compilation as soon as anyone calls any methods on CephContext, but that's just personal taste.

Reply:
You mean the forward declaration? In this case, there are a couple of ways forward. My personal preference is this: as part of these changes, nothing here is using that CephContext* parameter any more (and most weren't using it to begin with!). I'm totally fine getting rid of the backward-compatible functions that take it-- you can see that I've already re-implemented them all to ignore the CephContext* anyway, I don't think it will be a great effort to change the callsites in RGW. The question is, what about the callsites /not/ in RGW? (If any.)